experimental/sdk/webpubsub-socketio-extension/src/SIO/components/web-pubsub-adapter.ts

```diff
@@ -62,28 +63,28 @@ export class WebPubSubAdapterInternal extends NativeInMemoryAdapter {
    * @param packet - the packet object
    * @param opts - the options
    */
-  public override async broadcast(packet: Packet, opts: BroadcastOptions): Promise<void> {
-    debug(`broadcast packet ${JSON.stringify(packet)}`);
+  public override async broadcast(packet: SioPacket, opts: BroadcastOptions): Promise<void> {
+    debug(`broadcast, start, packet ${JSON.stringify(packet)}`);
+    packet.nsp = this.nsp.name;
 
     // Modified from https://github.com/socketio/socket.io-adapter/blob/2.5.2/lib/index.ts#L233
-    const encodedPackets = this._encoder.encode(packet);
-
+    // if packet owns binary attachements, `.encode` returns [string, ...buffers]. Otherwise, returns a single element of string.
+    let encodedPackets = this._encoder.encode(packet);
     const oDataFilter = this._buildODataFilter(opts.rooms, opts.except);
 
-    debug(`broadcast encodedPackets = ${encodedPackets}, oDataFilter = ${oDataFilter}`);
+    debug(`broadcast encodedPackets = "${encodedPackets}", oDataFilter = "${oDataFilter}"`);
 
-    const packets = Array.isArray(encodedPackets) ? encodedPackets : [encodedPackets];
-    const sendOptions = { filter: oDataFilter };
-    for (let encodedPacket of packets) {
-      if (typeof encodedPacket === "string") {
-        encodedPacket = "4" + encodedPacket.toString();
-        sendOptions["contentType"] = "text/plain";
-      } else {
-        sendOptions["contentType"] = "application/octet-stream";
-      }
+    // Ensure `encodedPackets` is an array whose binary attachements are base64-encoeded.
+    encodedPackets = Array.isArray(encodedPackets)
```

Comment:
Seems you don't need to handle it by hand, use `EioParser.encodePayload()` to encode. And extract this encoding from socketio packet to engineio payload string to a method as it needs to be reused by broadcastWithAck

Reply:
updated as suggested